### PR TITLE
log loss fix instability

### PR DESCRIPTION
**Pull request by ShvetsKS**

### CODE REVIEW

**Comment by ShvetsKS** [comment]
make test

**Comment by ShvetsKS** [comment]
precommit has passed
e9eaa193-3070-f103-8847-8cdcd4b723a1

File: algorithms/kernel/objective_function/logistic_loss/logistic_loss_dense_default_batch_impl.i [hunk 1]
```diff
@@ -336,27 +339,48 @@ services::Status LogLossKernel<algorithmFPType, method, cpu>::doCompute(const al
                 DAAL_CHECK_BLOCK_STATUS(gr);
                 g = gr.get();
             }
-            const size_t nRowsInBlock = (p < 5000 ? 5000/p : 1);
+
+            char trans = 'T';
+            char notrans = 'N';
+            algorithmFPType one = 1.0;
+            DAAL_INT yDim = 1;
+            DAAL_INT dim = (DAAL_INT)p;
```
Comment (averbukh) [inline algorithms/kernel/objective_function/logistic_loss/logistic_loss_dense_default_batch_impl.i]: Violates SDL

File: algorithms/kernel/objective_function/logistic_loss/logistic_loss_dense_default_batch_impl.i [hunk 2]
```diff
@@ -336,27 +339,48 @@ services::Status LogLossKernel<algorithmFPType, method, cpu>::doCompute(const al
...
+
+            const size_t nRowsInBlock = 1024;
             const size_t nDataBlocks = n / nRowsInBlock + !!(n%nRowsInBlock);
             const auto nThreads = daal::threader_get_threads_number();
-
             if((nThreads > 1) && (nDataBlocks > 1))
             {
-                TlsSum<algorithmFPType, cpu> tlsData(nBeta);
+                TArray<algorithmFPType, cpu> grads(nDataBlocks*p);
+                algorithmFPType* gradsPtr = grads.get();
```
Comment (averbukh) [inline algorithms/kernel/objective_function/logistic_loss/logistic_loss_dense_default_batch_impl.i]: algorithmFPType * const

File: algorithms/kernel/objective_function/logistic_loss/logistic_loss_dense_default_batch_impl.i [hunk 3]
```diff
@@ -336,27 +339,48 @@ services::Status LogLossKernel<algorithmFPType, method, cpu>::doCompute(const al
...
+                daal::services::internal::service_memset<algorithmFPType, cpu>(gradsPtr, algorithmFPType(0), nDataBlocks*p);
                 daal::threader_for(nDataBlocks, nDataBlocks, [&](size_t iBlock)
                 {
                     const size_t iStartRow = iBlock*nRowsInBlock;
                     const size_t nRowsToProcess = (iBlock == nDataBlocks - 1) ? n - iBlock * nRowsInBlock : nRowsInBlock;
                     const auto px = x + iStartRow * p;
-                    auto pg = tlsData.local();
+                    auto ps = s + iStartRow;
+                    const auto py = y + iStartRow;
+                    DAAL_INT nN   = (DAAL_INT)nRowsToProcess;
```
Comment (averbukh) [inline algorithms/kernel/objective_function/logistic_loss/logistic_loss_dense_default_batch_impl.i]: Violates SDL

File: algorithms/kernel/objective_function/logistic_loss/logistic_loss_dense_default_batch_impl.i [hunk 4]
```diff
@@ -336,27 +339,48 @@ services::Status LogLossKernel<algorithmFPType, method, cpu>::doCompute(const al
...
+                    PRAGMA_VECTOR_ALWAYS
                     for(size_t i = 0; i < nRowsToProcess; ++i)
                     {
-                        const algorithmFPType d = (s[iStartRow + i] - y[iStartRow + i]);
-                        for(size_t j = 0; j < p; ++j)
-                            pg[j + 1] += d * px[i * p + j];
+                        ps[i] -= py[i];
+                    }
+                    daal::internal::Blas<algorithmFPType, cpu>::xxgemm(&notrans, &trans, &yDim, &dim, &nN, &one, s + iStartRow, &yDim, px,
+                                                                                      &dim, &one, pg, &yDim);
```
Comment (averbukh) [inline algorithms/kernel/objective_function/logistic_loss/logistic_loss_dense_default_batch_impl.i]: Minor: Fix alignment

Comment (ShvetsKS) [reply algorithms/kernel/objective_function/logistic_loss/logistic_loss_dense_default_batch_impl.i]: alignment was already done as
const size_t nDataBlocks = n / nRowsInBlock + !!(n%nRowsInBlock);

File: algorithms/kernel/objective_function/logistic_loss/logistic_loss_dense_default_batch_impl.i (hunk 1 above)
Comment (ShvetsKS) [reply algorithms/kernel/objective_function/logistic_loss/logistic_loss_dense_default_batch_impl.i]: fixed

File: algorithms/kernel/objective_function/logistic_loss/logistic_loss_dense_default_batch_impl.i (hunk 3 above)
Comment (ShvetsKS) [reply algorithms/kernel/objective_function/logistic_loss/logistic_loss_dense_default_batch_impl.i]: fixed

**Comment by ShvetsKS** [comment]
new precommit:
e9eb5ba9-6cc5-f1e3-8556-8cdcd4b723a1

**Pull request merged**